### PR TITLE
Only grant or reject spillback lease request

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?
Status quo:
GCS based actor scheduling: For infeasible and unavailable actors, they are waited inside GCS until resource is available. For available actors, GCS picks the best raylet and sends lease request to it. The chosen raylet can either grant or reject the lease, but no spillback to other raylets.

Raylet based normal task scheduling: For infeasible tasks, they are waited inside the raylet until resource is available. For unavailable tasks, they are spilled back to the best feasible raylet and waited there. For available tasks, they are either granted locally or spilled back to other better raylets.

Problem:
Since raylet's local view of other raylets can be stale (stale until the next report), there can be lots of unnecessary spillbacks for a single task. Also each spillback updates local view of other raylets (i.e.  subtract cpu) even though they are likely to be redirected again, which cause the local view even more inaccurate. This is particular problematic when we launch many tasks in a short time (e.g. Dataset read tasks).  This is actually one of the problems GCS based actor scheduling solved: https://docs.google.com/document/d/1oPMqEj9YLCh75oArgN7XrViOQnoTdeTtg6QNRMMNVB0/edit#

Proposal:
While still doing distributed normal task scheduling, we can let the first raylet of the task act like a local gcs (i.e. only first raylet is responsible for finding the best raylet to run the task): For infeasible tasks, they are waited inside first raylet and for unavailable tasks, they are waited inside the best feasible raylet until resource is available. For available task, first raylet either grant the lease or spillback to the best raylet. The chosen raylet either grants the lease if the local available resources are enough or reject otherwise (can happen when the first raylet had a stale view of the chosen raylet), but no spillback to other raylets. If the lease is rejected, it will go back to the first raylet for rescheduling (hopefully the local view is refreshed).
Pros: Simpler to reason about the scheduling behavior since there is a single decision maker (i.e first raylet). Avoid the excessive spillbacks due to stale local resource view.
 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by jjyao** [comment]
One example of excessive spillbacks: https://docs.google.com/document/d/1nz68kjDvDi1ueHGVfo9O3lJ6imFR-OUnSHDb_mDnYNI/edit

**Comment by scv119** [comment]
This is a very clear description! can you add a bit clarification for "available/unavailable/infeasible" task?

**Comment by rkooo567** [comment]
Are there any concerns about head node raylet being overloaded? Also, can you write how the spillback works with this proposal in this doc? (For a comparison) https://docs.google.com/document/d/1nz68kjDvDi1ueHGVfo9O3lJ6imFR-OUnSHDb_mDnYNI/edit

**Comment by scv119** [comment]
I think the missing part of the proposal is _why_ it will work better. 

After discuss with @rkooo567  I have some guess, such as it reduce the likelihood of stale views, and it's almost identical to GCS based scheduling if all tasks are submitted by driver.

**Comment by jjyao** [comment]
> This is a very clear description! for completeness, can you add a bit clarification for "available/unavailable/infeasible" task?

I think I probably didn't use the right terms:

available task means that there is available free resources and the task can be scheduled now.
unavailable task means that there is not enough free resources due to other running tasks but it's feasible without adding more resources to the cluster.
infeasible task means the task cannot run without adding more resources to the cluster.

**Comment by jjyao** [comment]
Included proper round robin fix in this PR as well. Also changed dataset test to show that we can avoid `_spread_resource_prefix` (although the test is just a toy example). Going to test it out with real ML ingest workload.

**Comment by rkooo567** [comment]
> the first lease request for a task needs to go to the home raylet anyway

So, the overhead is the same only until the first spillback, and after that it increases on a head raylet? Since previously, scheduling decisions were done from remote nodes whereas now everything is done from the head node raylet. I think "grant or lease" decision is really cheap compared to the scheduling overhead (only need to compare local resources to the task request vs comparing task requests to all nodes resources). 

I feel like in this case, isn't decentralized scheduling is basically not working? (all scheuduling work is done at the home raylet, and other raylets only check if it is scheduleable locally or not, which is probably very cheap)?

**Comment by ericl** [comment]
> So, the overhead is the same only until the first spillback, and after that it increases on a head raylet?

In practice the home raylet is doing the vast majority of the work. You only get multiple spillbacks on rare edge conditions, so this will not appreciably change load.

It was never intended that subsequent raylets make _policy decisions_ again, that's a flat out bug.

**Comment by jjyao** [comment]
All the existing tests are passing. Would like to get the review started while I'm adding comments and tests.

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (ericl) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:253]: Can we RAY_CHECK later that best_node_id is local node if grant_or_reject is true.

Consider just immediately returning the local node id if feasible here instead of going through the full policy logic.

File: src/ray/raylet/node_manager.cc
Comment (ericl) [inline src/ray/raylet/node_manager.cc:1601]: Why don't we follow this for normal tasks as well? Seems that would simplify the logic

File: src/ray/raylet/node_manager.cc
Comment (ericl) [inline src/ray/raylet/node_manager.cc:1632]: How about avoiding calling QueueAndSchedule for grant_or_reject tasks?

File: src/ray/raylet/node_manager.cc
Comment (jjyao) [reply src/ray/raylet/node_manager.cc:1601]: I don’t follow this to make it explicit that for normal task, we don’t use that normal task resources feedback loop. That only used by gcs actor scheduling.

File: src/ray/raylet/node_manager.cc
Comment (jjyao) [reply src/ray/raylet/node_manager.cc:1632]: We still need to do normal QueueAndSchedule since we still need things like resolving/waiting args.

**Comment by jjyao** [comment]
For reading 200 files in a 40 nodes cluster, sometimes it still cannot achieve perfect spreading due to the grpc resource broadcast which basically reset the more accurate local view of other raylets.

**Comment by jjyao** [comment]
I still need to resolve merge conflicts, so don't merge it yet.

**Comment by jjyao** [comment]
Passed core nightly: https://buildkite.com/ray-project/periodic-ci/builds/1554

File: src/ray/protobuf/node_manager.proto
Comment (stephanie-wang) [inline src/ray/protobuf/node_manager.proto:48]:
```suggestion
  // locally schedulable or reject the request. Else, the raylet may return another raylet at which to retry the request.
```

File: src/ray/raylet/node_manager.cc
Comment (stephanie-wang) [reply src/ray/raylet/node_manager.cc:1601]: Hmm I agree with @ericl here. Would it work if we changed this to be agnostic to actor creation tasks vs normal tasks? I think we just need two flags from the client here:

- grant_or_reject (always true for GCS)
- include_current_resources (always true for GCS to update resources view)

My higher level question here is if there's a known issue if we queue actor creation tasks instead of rejecting immediately?

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (stephanie-wang) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:173]: Thanks for the cleanup :)

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (stephanie-wang) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:246]: Can you add a comment here to explain?

File: src/ray/raylet/scheduling/cluster_task_manager.cc
Comment (stephanie-wang) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:1054]: I think it might be simpler if we don't pass the grant_or_reject flag to the ClusterResourceScheduler and just decide which ClusterResourceScheduler methods to call from this class. That way:
1. This class can control when we actually need to call the full policy.
2. We have to check the flag in this class anyway, so this way we keep the flag contained to this class.

File: src/ray/raylet/scheduling/scheduling_policy.cc
Comment (stephanie-wang) [inline src/ray/raylet/scheduling/scheduling_policy.cc:90]: Can we use the saved index no matter what the spread threshold is?

File: src/ray/raylet/scheduling/scheduling_policy.cc
Comment (stephanie-wang) [inline src/ray/raylet/scheduling/scheduling_policy.cc:92]: Won't this violate the previous behavior where we check the local node first?

File: src/ray/raylet/scheduling/scheduling_policy.cc
Comment (stephanie-wang) [inline src/ray/raylet/scheduling/scheduling_policy.cc:46]: Do you think it would be better to always iterate through `nodes_` instead of building this other copy? I find this structure a bit weird anyway, and it would make more sense now to try to keep the saved index consistent.

**Review by stephanie-wang — DISMISSED** [review]
I have a couple questions about high-level behavior:

1. How should the new policy behave when a raylet is chosen through data locality?
2. Will this still follow the old policy where we try to pick the local node first?

Also, can we add unit tests for these if they don't already exist?

**Comment by stephanie-wang** [comment]
Also, I think it would be better if you could split out the fix for round-robin scheduling into its own PR. Seems like it's not really related to the grant/reject spillback change.

**Comment by jjyao** [comment]
> I have a couple questions about high-level behavior:
> 
> 1. How should the new policy behave when a raylet is chosen through data locality?

Yea, the raylet chosen through data locality is treated the same as home raylet. Actually what I mean by home raylet is whatever raylet returned from `lease_policy_->GetBestNodeForTask`

> 2. Will this still follow the old policy where we try to pick the local node first?

Yea, if it's not spread scheduling, then local node is still considered first.

> 
> Also, can we add unit tests for these if they don't already exist?

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (jjyao) [reply src/ray/core_worker/transport/direct_task_transport.cc:519]: Yea, it's set to false if provided `raylet_address` is nullptr and we choose the first raylet based on data locality below.

**Comment by jjyao** [comment]
> Also, I think it would be better if you could split out the fix for round-robin scheduling into its own PR. Seems like it's not really related to the grant/reject spillback change.

Yea, I can do that. I was initially having them together to make sure they work nicely with each other.

File: src/ray/raylet/scheduling/scheduling_policy.cc
Comment (jjyao) [reply src/ray/raylet/scheduling/scheduling_policy.cc:90]: Will do it in the separate round robin pr.

File: src/ray/raylet/scheduling/scheduling_policy.cc
Comment (jjyao) [reply src/ray/raylet/scheduling/scheduling_policy.cc:92]: Will do it in the separate round robin pr.

File: src/ray/raylet/scheduling/scheduling_policy.cc
Comment (jjyao) [reply src/ray/raylet/scheduling/scheduling_policy.cc:46]: Will do it in the separate round robin pr.

File: src/ray/raylet/scheduling/cluster_task_manager.cc
Comment (jjyao) [reply src/ray/raylet/scheduling/cluster_task_manager.cc:1054]: Actually after a second thought, I think it's probably still better to have it inside ClusterResourceScheduler since it's about scheduling and ClusterResourceScheduler should be responsible for choosing what scheduling policy to use based on the input scheduling constraints (e.g force_spillback, grant_or_reject) and expose a single GetBestSchedulableNode API. What do you think?

File: src/ray/raylet/node_manager.cc
Comment (jjyao) [reply src/ray/raylet/node_manager.cc:1601]: Going to always queue actor creation tasks as well to be consistent with normal tasks. We can still reject once they are queued.

**Comment by jjyao** [comment]
Move round robin related changes to a separate pr.

File: src/ray/raylet/scheduling/cluster_task_manager.cc
Comment (stephanie-wang) [reply src/ray/raylet/scheduling/cluster_task_manager.cc:1054]: Hmm I still think it's nicer to have it in the ClusterTaskManager (which is also about scheduling, maybe this is part of the issue), so that we can keep the logic more modular. So for example, now we have to test both ClusterResourceScheduler and ClusterTaskManager logic to make sure grant_or_reject is respected, whereas if we kept it only in ClusterTaskManager, we would only have to test that class.

But I think it's pretty minor, so I'll leave it up to you.

File: src/ray/core_worker/test/direct_task_transport_test.cc
Comment (stephanie-wang) [inline src/ray/core_worker/test/direct_task_transport_test.cc:181]: Can we add a test to check that `grant_or_reject` is set properly by the client?

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_task_transport.cc:563]:
```suggestion
            // Retry the request at the home raylet since the resource view may be refreshed.
```

File: src/ray/raylet/node_manager.cc
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc:1588]:
```suggestion
    cluster_task_manager_->QueueAndScheduleTask(task, /*grant_or_reject=*/false, reply, send_reply_callback);
```

File: src/ray/raylet/node_manager.cc
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc:1602]:
```suggestion
    // for GCS based actor scheduling, return a rejection
```
etc.

File: src/ray/raylet/node_manager.cc
Comment (stephanie-wang) [reply src/ray/raylet/node_manager.cc:1625]: Could probably simplify this further by using the same `send_reply_callback_wrapper` for both grant_or_reject = true/false. Right now, looks like we're calling `send_reply_callback` in 3 different places and it'd be nice to get this down to one for clarity.

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (stephanie-wang) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:963]: Is this change related?

File: src/ray/raylet/scheduling/cluster_resource_scheduler_test.cc
Comment (stephanie-wang) [inline src/ray/raylet/scheduling/cluster_resource_scheduler_test.cc:339]: Can you add a test for a case where the policy should return a different node if grant_or_reject=true vs false?

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (stephanie-wang) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:246]: Discussed offline: to simplify, we can just reject the task immediately if it is not locally schedulable. This way, the policy will always be to queue at the home raylet until remote resources become available.

**Review by stephanie-wang — COMMENTED** [review]
It looks good, just some suggestions for cleanups.

File: src/ray/raylet/scheduling/cluster_task_manager.cc
Comment (jjyao) [reply src/ray/raylet/scheduling/cluster_task_manager.cc:1054]: Ok, Thinking ClusterTaskManager is also about scheduling makes more sense to me now. Will change it.

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (jjyao) [reply src/ray/raylet/scheduling/cluster_resource_scheduler.cc:963]: This is #20026, but it's reverted because somehow it made 1tb autoscaler shuffling timed out. But when it's combined with this PR it worked so I just redo it here since it's small.

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (jjyao) [reply src/ray/raylet/scheduling/cluster_resource_scheduler.cc:246]: Discussed offline: will keep the behavior as it is for now and discuss what the better behavior is later.

File: src/ray/raylet/node_manager.cc
Comment (jjyao) [reply src/ray/raylet/node_manager.cc:1625]: Good point. Changed to always use `send_reply_callback_wrapper`

**Review by ericl — COMMENTED** [review]
Also LGTM, but wondering if we should trigger nightly tests on it, given the sensitivity of the change.

**Comment by jjyao** [comment]
Core nightly passed: https://buildkite.com/ray-project/periodic-ci/builds/1630#_

**Review by ericl — APPROVED** [review]

**Pull request merged**
